File: chapter1/004/main.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	fmt.Println(palindromePermutation("Tact Coa"))
	fmt.Println(palindromePermutation("arara"))
	fmt.Println(palindromePermutation("Abba"))
	fmt.Println(palindromePermutation("obo"))

	fmt.Println(palindromePermutation(""))
	fmt.Println(palindromePermutation("a"))

	fmt.Println(palindromePermutation("apex"))
	fmt.Println(palindromePermutation("stenio"))
	fmt.Println(palindromePermutation("concatenar"))
	fmt.Println(palindromePermutation("pararalelepipedo"))
}

func palindromePermutation(s string) bool {
	s = strings.Replace(s, " ", "", -1)
	s = strings.ToLower(s)

	gates := make(map[rune]bool)

	for _, r := range s {
		gates[r] = !gates[r]
	}

	open := 0
	for _, v := range gates {
		if v {
			open++
		}
	}

	if open > 1 {
		return false
	}

	return true
}

File: chapter1/004/main.py
def main():
	print(is_palindrome_permutation("Tact Coa"))
	print(is_palindrome_permutation("arara"))
	print(is_palindrome_permutation("Abba"))
	print(is_palindrome_permutation("obo"))
	print(is_palindrome_permutation(""))
	print(is_palindrome_permutation("a"))

	print(is_palindrome_permutation("apex"))
	print(is_palindrome_permutation("stenio"))
	print(is_palindrome_permutation("concatenar"))
	print(is_palindrome_permutation("pararalelepipedo"))

def is_palindrome_permutation(string):
    string = string.lower()
    string = string.replace(" ", "")

    gates = {}
    for c in string:
        gates[c] = not gates.get(c, False)
    
    opens = 0
    for gate_open in gates.values():
        if gate_open:
            opens += 1
    
    if opens > 1:
        return False
    
    return True

main()